fs/dafsserver/dafsserver.cpp

```diff
@@ -2934,68 +2949,87 @@ class CRemoteFileServer : implements IRemoteFileServer, public CInterface
             touch();
             try
             {
-                if (!gotSize)
+                while (true)
                 {
-                    // left represents amount we have read of leading size32_t (normally expect to be read in 1 go)
-                    if (0 == msg.length()) // 1st time
-                        msgWritePtr = (byte *)msg.reserveTruncate(sizeof(size32_t));
-                    size32_t szRead;
-                    sock->read(msgWritePtr, 1, sizeof(size32_t)-left, szRead);
-                    left += szRead;
-                    msgWritePtr += szRead;
-                    if (left == sizeof(size32_t))
+                    if (!gotSize)
                     {
-                        gotSize = true;
-                        msg.read(left);
-                        msg.clear();
-                        try
-                        {
-                            msgWritePtr = (byte *)msg.reserveTruncate(left);
-                        }
-                        catch (IException *e)
+                        // left represents amount we have read of leading size32_t (normally expect to be read in 1 go)
+                        if (0 == msg.length()) // 1st time
+                            msgWritePtr = (byte *)msg.reserveTruncate(sizeof(size32_t));
+                        size32_t szRead;
+                        if (readtmsCheckGC(sock, msgWritePtr, 0, sizeof(size32_t)-left, szRead, WAIT_FOREVER))
+                            THROWJSOCKEXCEPTION(JSOCKERR_graceful_close);
```

Comment:
Is this different than a sock->read(msgWritePtr, 0, sizeof(size32_t)-left, szRead, WAIT_FOREVER, false); ?

Reply:
no, it's the same. I'll change it.
i.e. if hit GC before then hit max then we haven't receive the packet size and we want to throw GC exception.